src/app/website/posts/zkp.markdown

```diff
@@ -0,0 +1,83 @@
+
+* If you are cheating (i.e., if you do not know a 3-coloring), then Verifier should be able to catch you -- this property is called  _soundness_
+
+* Verifier should not learn anything about the 3-coloring -- this property is called  _zero knowledgeness_
+
+Now, let's try to construct a ZKP protocol. Imagine the graph drawn on the floor in a closed space. Per the 3-coloring you know, you would place the corresponding colored balls on the nodes. Then you would completely cover the balls with inverted opaque bowls. 
+Then, Verifier comes and points at an edge of her choice. You would lift up the two bowls on either side of the edge. Verifier verifies that the balls revealed are of different colors. If they are not, you are caught cheating. 
+
+Notice that if you didnt really know a 3-coloring (meaning there is at least one edge with same colored balls on either side), you would get caught with some probability (namely, at least, $1/E$ where $E$ is the number of edges in the graph). Also, no further information about the 3-coloring is revealed to Verifier, since whatever Verifier viewed could be simulated by just picking two random, but differently-colored balls. Thus, this protocol provides some degree of soundness and perfect zero knowledgeness. 
+
+## Soundness
+
+Note that there is still some significant probability that you could cheat and still get away with it (namely, $(1 - 1/E)$). This probability is called the _soundness error_.  We would like to reduce this error to negligible. Here is an idea.  Repeat the above protocol multiple times. Now, you can get away only if you get away in each one of those executions. This significantly reduces the soundness error, as quantified in the following.   
+
+
+
+
+
+
+### More Rigorously,…
+
+The more rounds you execute, higher is the Verifier’s confidence on the soundness of your claim. Let’s say that you do not know a 3-coloring for the graph.  The probability that you will not get caught can be bounded as follows. 
+
+Let $N$ be the total number of rounds. 
+<div class="katex-block">
+```
+\Pr[\text{You will not be caught in Round }i \leq N] \leq 1 - \frac{1}{E}
+```
+</div>
+
+<div class="katex-block">
+```
+\Pr[\text{You will not be caught in any of the N rounds}] \leq \left(1 - \frac{1}{E} \right)^N
+```
+</div>
+
+
+
+
+
+
+
+ 
+
+## Zero Knowledgeness
+
+Unfortunately, there is an issue in the above protocol. Since Verifier gets to see the coloring two nodes at a time, she can learn the entire 3-coloring by running enough rounds. Here is an idea on how we can get around this issue. After every round, you would ask Verifier to step out of sight, you would randomly permute the colors and again cover all the nodes. That way, anything that Verifier might have learned in one round is not relevant in the subsequent rounds, since whatever Verifier viewed can be simulated by just picking two random, but differently-colored balls in each round. 
```

Comment:
> Here is an idea on how we can get around this issue.

`Luckily, we can get around this issue:`

Reply:
Done.